### PR TITLE
IntegrationMapper.ConvertType simplification and Ducktype optimization

### PR DESCRIPTION
**Pull request by tonyredondo**
## Summary of changes

This PR:

- Simplifies the `IntegrationMapper.ConvertType()` method.
- Adds `ref TReturn GetInternalDuckTypedInstance<TReturn>()` in `IDuckType` interface allowing to extract the internal instance from the proxy without boxing or casting as the `Instance` property.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (5216)](https://github.com/DataDog/dd-trace-dotnet/tree/71c127a355d1dcf9e517502fca338127bc437a18)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/08c9c4cda7ea7e4b53fe7e1e8d194fe2a0164769)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5216) - mean (74ms)  : 63, 84
     .   : milestone, 74,
    master - mean (75ms)  : 65, 86
     .   : milestone, 75,

    section CallTarget+Inlining+NGEN
    This PR (5216) - mean (984ms)  : 963, 1006
     .   : milestone, 984,
    master - mean (989ms)  : 969, 1009
     .   : milestone, 989,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5216) - mean (110ms)  : 108, 112
     .   : milestone, 110,
    master - mean (111ms)  : 107, 115
     .   : milestone, 111,

    section CallTarget+Inlining+NGEN
    This PR (5216) - mean (717ms)  : 692, 742
     .   : milestone, 717,
    master - mean (710ms)  : 691, 728
     .   : milestone, 710,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5216) - mean (95ms)  : 90, 100
     .   : milestone, 95,
    master - mean (93ms)  : 90, 96
     .   : milestone, 93,

    section CallTarget+Inlining+NGEN
    This PR (5216) - mean (673ms)  : 650, 696
     .   : milestone, 673,
    master - mean (671ms)  : 646, 696
     .   : milestone, 671,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5216) - mean (188ms)  : 185, 191
     .   : milestone, 188,
    master - mean (188ms)  : 184, 191
     .   : milestone, 188,

    section CallTarget+Inlining+NGEN
    This PR (5216) - mean (1,062ms)  : 1040, 1084
     .   : milestone, 1062,
    master - mean (1,057ms)  : 1033, 1082
     .   : milestone, 1057,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5216) - mean (271ms)  : 266, 275
     .   : milestone, 271,
    master - mean (271ms)  : 266, 275
     .   : milestone, 271,

    section CallTarget+Inlining+NGEN
    This PR (5216) - mean (1,063ms)  : 1043, 1084
     .   : milestone, 1063,
    master - mean (1,055ms)  : 1024, 1085
     .   : milestone, 1055,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5216) - mean (260ms)  : 257, 263
     .   : milestone, 260,
    master - mean (259ms)  : 253, 264
     .   : milestone, 259,

    section CallTarget+Inlining+NGEN
    This PR (5216) - mean (1,001ms)  : 983, 1018
     .   : milestone, 1001,
    master - mean (994ms)  : 967, 1022
     .   : milestone, 994,

```

File: tracer/test/Datadog.Trace.DuckTyping.Tests/Fields/ReferenceType/ReferenceTypeFieldTests.cs [hunk 1]
```diff
@@ -50,6 +50,9 @@ public void StaticReadonlyFields(string obscureObjectName)
             var duckAbstract = obscureObject.DuckCast<ObscureDuckTypeAbstractClass>();
             var duckVirtual = obscureObject.DuckCast<ObscureDuckTypeVirtualClass>();
 
+            ref var originalObject = ref ((IDuckType)duckInterface).GetInternalDuckTypedInstance<object>();
+            Assert.NotNull(originalObject);
```
Comment (andrewlock) [inline tracer/test/Datadog.Trace.DuckTyping.Tests/Fields/ReferenceType/ReferenceTypeFieldTests.cs:54]: You can assert more than that, right? Can't we assert that it _is_ `obscureObject`?

File: tracer/src/Datadog.Trace/DuckTyping/DuckType.cs [hunk 2]
```diff
@@ -447,7 +447,7 @@ private static FieldInfo CreateIDuckTypeImplementation(TypeBuilder proxyTypeBuil
             il.Emit(OpCodes.Ret);
             propInstance.SetGetMethod(getPropInstance);
 
-            PropertyBuilder propType = proxyTypeBuilder.DefineProperty("Type", PropertyAttributes.None, typeof(Type), null);
+            PropertyBuilder propType = proxyTypeBuilder.DefineProperty(nameof(IDuckType.Type), PropertyAttributes.None, typeof(Type), null);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/DuckTyping/DuckType.cs:450]: nice

**Review by andrewlock — APPROVED** [review]
Nice

**Review by kevingosse — APPROVED** [review]

File: tracer/test/Datadog.Trace.DuckTyping.Tests/Fields/ReferenceType/ReferenceTypeFieldTests.cs (hunk 1 above)
Comment (tonyredondo) [reply tracer/test/Datadog.Trace.DuckTyping.Tests/Fields/ReferenceType/ReferenceTypeFieldTests.cs]: true! done in [7477206](https://github.com/DataDog/dd-trace-dotnet/pull/5216/commits/747720610962dbe522b074ffb123f7b407ebb08a)

**Pull request merged**